tonic/datasets/ncars.py

```diff
@@ -0,0 +1,85 @@
+import os
+from tonic.io import read_aedat4 
+import numpy as np
+from tonic.dataset import Dataset
+from tonic.download_utils import extract_archive
+
+
+class NCARS(Dataset):
+    """N-Cars dataset <https://www.prophesee.ai/dataset-n-cars-download/>. Events have (txyp) ordering.
+    ::
+        @inproceedings{sironi2018hats,
+          title={HATS: Histograms of averaged time surfaces for robust event-based object classification},
+          author={Sironi, Amos and Brambilla, Manuele and Bourdis, Nicolas and Lagorce, Xavier and Benosman, Ryad},
+          booktitle={Proceedings of the IEEE Conference on Computer Vision and Pattern Recognition},
+          pages={1731--1740},
+          year={2018}
+        }
+    Parameters:
+        save_to (string): Location to save files to on disk.
+        train (bool): If True, uses training subset, otherwise testing subset.
+        transform (callable, optional): A callable of transforms to apply to the data.
+        target_transform (callable, optional): A callable of transforms to apply to the targets/labels.
+    """
+
+    url = "http://www.prophesee.ai/resources/Prophesee_Dataset_n_cars.zip"
+    filename = "Prophesee_Dataset_n_cars.zip"
+    train_filename = "n-cars_train.zip"
+    test_filename = "n-cars_test.zip"
+    file_md5 = "553ce464d6e5e617b3c21ce27c19368e"
+    classes = ["background", "car"]
+
+    class_dict = {"background": 0, "cars": 1}
+
+    sensor_size = None  # different for every recording
+    minimum_y_value = 140
+    dtype = np.dtype([("t", np.int64), ("x", np.int16), ("y", np.int16), ("p", np.int8)])
+    ordering = dtype.names
+
+    def __init__(self, save_to, train=True, transform=None, target_transform=None):
+        super(NCARS, self).__init__(
+            save_to, transform=transform, target_transform=target_transform
+        )
+
+        if train:
+            self.folder_name = "train"
+        else:
+            self.folder_name = "test"
+
+        if not self._check_exists():
+            self.download()
+            extract_archive(os.path.join(self.location_on_system, self.train_filename))
+            extract_archive(os.path.join(self.location_on_system, self.test_filename))
+
+        file_path = os.path.join(self.location_on_system, self.folder_name)
+        for path, dirs, files in os.walk(file_path):
+            dirs.sort()
+            for file in files:
+                if file.endswith("dat"):
+                    self.data.append(path + "/" + file)
+                    self.targets.append(self.class_dict[os.path.basename(path)])
+
+    def __getitem__(self, index):
+        """
+        Returns:
+            a tuple of (events, target) where target is the index of the target class.
+        """
+        events = read_aedat4(self.data[index])# loris.read_file(self.data[index])["events"]
+        events = np.lib.recfunctions.rename_fields(events, {'ts': 't', 'is_increase': 'p'})
+        events = events.astype(self.dtype)
```

Comment:
does this line actually work? because we already have a structured array at this point coming from read_aedat4

Reply:
Yes, the array is already structured but I was trying to cast the data from unsigned to signed, as we did in CIFAR10DVS, to avoid overflow when performing operations on data. The code I was trying to implement is the following:
```
IIn [15]: a = np.array([(1, 2)], dtype=([('t', np.uint16), ('p', np.uint8)]))

In [16]: b = a.astype([('t', np.int16), ('p', np.int8)])

In [17]: a
Out[17]: array([(1, 2)], dtype=[('t', '<u2'), ('p', 'u1')])

In [18]: b
Out[18]: array([(1, 2)], dtype=[('t', '<i2'), ('p', 'i1')])
```
Do you have any suggestions?